src/main.cpp

```diff
@@ -168,9 +203,77 @@ namespace
         catch (const std::exception& e) {
             log_re::error({{"log_message", e.what()}, {"rule_engine_plugin", "metadata_guard"}});
         }
+        return CODE(RULE_ENGINE_CONTINUE);
+    }
+
+    auto exec_rule_atomic_metadata(std::list<boost::any>& _rule_arguments, irods::callback _effect_handler) -> irods::error 
+    {
+        auto is_input_valid = [](const bytesBuf_t* _input) -> std::tuple<bool, std::string>
+        {
+            if (!_input) {
+                return {false, "Missing JSON input"};
+            }
+
+            if (_input->len <= 0) {
+                return {false, "Length of buffer must be greater than zero"};
+            }
+
+            if (!_input->buf) {
+                return {false, "Missing input buffer"};
+            }
 
+            return {true, ""};
+        };
+
+        try {
+            auto* input_bb = boost::any_cast<bytesBuf_t*>(*std::next(std::begin(_rule_arguments), 2));
+            auto& rei = get_rei(_effect_handler);
+            const auto config = load_plugin_config(rei);
+
+
+            if (const auto [valid, msg] = is_input_valid(input_bb); !valid) {
+                log_re::error({{"log_message", msg}});
+                return ERROR(INPUT_ARG_NOT_WELL_FORMED_ERR, msg);
+            }
+
+            json input = json::parse(std::string(static_cast<const char*>(input_bb->buf), input_bb->len));
+
+            const auto& operations = input.at("operations");
+            std::vector<irods::error> op_codes; 
+
+            std::transform(operations.begin(), operations.end(), std::back_inserter(op_codes), [&rei](json op) {
+                return rule_logic(op.at("attribute").get_ref<const std::string&>(), rei);
+            });
```

Comment:
This processes ALL elements in the JSON object.

It would be better to return from the plugin as soon as we detect a violation.